File: dsa_leetcode_beginner/0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
#include <unordered_map>
#include <string>
#include <algorithm>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> charIndexMap;
        int maxLength = 0, left = 0;

        for (int right = 0; right < s.length(); ++right) {
            char currentChar = s[right];
            if (charIndexMap.find(currentChar) != charIndexMap.end() && charIndexMap[currentChar] >= left) {
                left = charIndexMap[currentChar] + 1;
            }
            charIndexMap[currentChar] = right;
            maxLength = max(maxLength, right - left + 1);
        }

        return maxLength;
    }
};

File: dsa_leetcode_beginner/0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
                
        # Initialize the pointers and the set
        start, end, maxLength = 0, 0, 0
        charSet = set()
        
        while end < len(s):
            if s[end] not in charSet:
                charSet.add(s[end])
                maxLength = max(maxLength, end - start + 1)
                end += 1
            else:
                charSet.remove(s[start])
                start += 1
                
        return maxLength


#Question: https://leetcode.com/problems/longest-substring-without-repeating-characters
#Blog: https://blog.unwiredlearning.com/longest-substring-without-repeating-characters